### DISCUSSION TITLE
OIDC client Redirect URIs validation

### DISCUSSION THREAD

**Discussion by cgeorgilakis** [keycloak/keycloak#12376] [Q&A]
Keycloak redirect URIs in OIDC client can be without schema. Denylist schemas are only 'data' and 'javascript'. 

[OIDC Authentication Request specification](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest) states about redirect_uri:

> Redirection URI to which the response will be sent. This URI MUST exactly match one of the Redirection URI values for the Client pre-registered at the OpenID Provider, with the matching performed as described in Section 6.2.1 of [[RFC3986]](https://openid.net/specs/openid-connect-core-1_0.html#RFC3986) (Simple String Comparison). When using this flow, the Redirection URI SHOULD use the https scheme; however, it MAY use the http scheme, provided that the Client Type is confidential, as defined in Section 2.1 of OAuth 2.0, and provided the OP allows the use of http Redirection URIs in this case. The Redirection URI MAY use an alternate scheme, such as one that is intended to identify a callback into a native application.

This is enhanced by [OAuth 2.0 Security Best Current Practice](https://tools.ietf.org/id/draft-ietf-oauth-security-topics-06.html).

Taking all these into account we want to ask : 

- Why does Keycloak accept redirect URIs without schema?
- Why do you accept wildcards in redirect URIs? This is risky - bad practice. See [Insufficient redirect URI validation](https://tools.ietf.org/id/draft-ietf-oauth-security-topics-06.html#rfc.section.3.1). Could at least this be a realm configuration ( do not accept wildcards for this realm)?
- Do you consider supporting two client categories ( web and native)? As you can see from OIDC specification and best practices web redirect URIs are more strict in validation.

**Comment by darius-m** [comment]
I am not entirely familiar with the standard and best practices, but you should be able to enforce restrictions through [client policies](https://www.keycloak.org/docs/latest/server_admin/index.html#_client_policies). At the least, not using wildcard redirect URIs is explicitly mentioned as something you can enforce using policies.

As to why the server supports it, I would argue that unless it is specifically against standard, the server should allow features in order to be compliant. If an administrator for whatever reason needs some functionality, they should be able to use Keycloak even if it would go against good practices (again, good practices are not part of the RFC). The service should allow people to decide how to use their systems.

This being said, I think the points you mentioned are valid. If possible, a stricter default configuration (e.g., implemented as a toggle, or some generally recommended policies) would be greatly appreciated, since administrators are also human and can make mistakes / forget some configuration.
